### ISSUE TITLE
Service startup ordering

### ISSUE THREAD

**Issue by JX7P** [ServiceManager/ServiceManager#4]
A seperate daemon ought to exist which will communicate with the master restarter and the repository manager to initialise services in the right order.

**Comment by uselessd** [comment]
This is going to be a very intriguing problem to solve.

No process managers really implement dependencies in much of the same way. The nature of the problem domain means that modeling dependencies between processes is intrinsically difficult because they don't have any bindings or relations in the traditional sense, but largely only to resources exported by other processes. 

The most common form of dependency system is to simply parse directives for calculating an ordering sequence. This is fine, but also rather weak and might as well be done manually - it's just before/after. These dependencies can be much more elegantly modeled as preconditions and postconditions, instead [1]. Have a need program that checks for something being up and is continuously restarted until it succeeds, and for conflicting dependencies or binding multiple services, just issue a command to kill a service in another service's Exec[Start|Stop][Pre|Post].

Transactional dependency systems are much tougher. SMF uses a barely readable internal graph engine, whereas systemd... you start a unit file, which allocates a unit, which queues a job, which runs a service unit, which joins a transaction, which is checked for consistency, which is then joined with an execution environment chain, which is then hopefully started. It's all voodoo heuristics and creates reliability problems like dependency loops and stuck jobs.

Something in between needs to be implemented. But, in general, I'd say simply being able to apply pre/postconditions to resources for availability is good enough.

[1] http://smarden.org/pape/djb/daemontools/noinit.html

**Comment by mkfifo** [comment]
If we can decide on how to encode the ordering dependencies and how such a daemon should communicate then I would be interested in helping write such a daemon to decide on an init ordering (a total ordering engine is straightforward to write in the non-parallel case).

Input:
A depends on B
A depends on C
B depends on C
C has no depends

This gives an ordering (time is from top to bottom): 

````
    C
    |
    B
    |
    A
```

Thus the order things must be started in is: C, B and then finally A.

It also depends on how much parallelism you want in initialising services, as in the above it is clear that (assuming they are hard dependencies) then you must start them in a linear (non parallel) order.

However if you had:
V depends on X
W depends on Y
X depends on Z
Y depends on Z

This gives an ordering (time is from top to bottom):

```
        Z
      /   \
     X     Y
     |     |
     V     W
```

1) start Z
2) once Z is started: you may now start X and Y in parallel
3.a) once X is started you may not start V
3.b) once Y is started you may now start W

**Comment by mkfifo** [comment]
There is also the case of processes dying

If service A depends on service B and both start cleanly

Some time later service B crashes and is restarted, do we need to restart service A? how do we know? what should the default be?

**Comment by uselessd** [comment]
Simplest case is service B registers a postcondition that A must be restarted along with it (e.g. in a post-crash file or PostCrash directive).

The problem with the above dependency proposal is it assumes all dependencies are hard, and furthermore that we are depending on a process. This I think is a misnomer. One typically depends on resources exported by a process, but to encode it as a hard dependency is gratuitous complication and mismatch over a more straightforward ordering.

Something like rcorder(8) or SMF is in need of being studied. Though, if we are to take the resource-driven aspect seriously, one could implement a system whereby services register for availability of a certain resource and are deferred from execution until the service manager's event loop is notified of the resource being up. A more generic way of the traditional inetd or "socket activation" style of lazy loading.

Parallelism is a red herring, especially with I/O-bound processes and daemon initialization times. Asynchronous boot is straightforward to implement.

**Comment by mkfifo** [comment]
I am unsure about "Simplest case is service B registers a postcondition that A must be restarted along with it "
As this requires that service B is aware of all it's dependencies that may need restarting (unless it has a blanket 'restart all services that depend on me', which would be fine).
From a distribution point of view service B and A may be in different packages, and service A may not always be installed, thus I think it makes more sense for service A to register that it needs to be restarted when B dies (but this distinction is potentially getting too finicky / picky).


The earlier suggestion was similar to `tsort(1)` (+ a parser), what is done with that information (whether deps. are hard or soft) is less interesting.


As for services vs resources if you change every 'service X' to 'resource X' doesn't most of the above remain unchanged?


`rcorder(8)` is very nice, and I quite like the breakdown by resource rather than service (requires and provide are both very nice primitives), although the require primitive is just 'this service must be placed before me in the ordering' and NOT 'this service must be completely started before me' (a hard dependency), so I presume we would need an additional primitive.

I am not familiar with how SMF handles things, which resource would you recommend reading?



I am also not convinced that parallelism is a red-herring, but leaving it out does make the whole system simpler so I am not overly fussed.

**Comment by uselessd** [comment]
The preconditions and postconditions would of course all be set up for the relevant services. This does place some complexity on the user, but it is not any more egregious than e.g. specifying systemd unit dependencies, which are ultimately fed into a transaction and queuing engine to produce an unobservable dependency graph in a non-deterministic fashion due to that in addition to parallelism and "socket activation". Here it would be explicit.

We're assuming a system where all services reflect installed packages, not just a bundle of unit files that have been overlayed from various sources.

When it comes to processes, the distinction between ordered and started before something is of little significance. It's all an eager execution graph, modulo asynchronicity and any laziness introduced by holding sockets or deferring upon resource availability.

The key to resources is that they're much more versatile. Dependencies tend to be static, crufty and try to bolt on a model whereby it is pretended service dependencies are as clear cut as e.g. library dependencies. Library dependencies are usually absolute, i.e. you need an object that publicly exports symbols matching an expected interface. Service dependencies are not, and in fact nearly all service dependencies are reducible to ordering problems.

A resource-driven approach would involve registering asynchronous notifiers for watching events on all sorts of names in a system, which would make it a generic deferred/conditional loader. Services are dispatched as their resources become available, which may not necessarily be tied to one particular daemon (I suppose a "provides" relationship could address this).

We then start to think about relationships with data. As, ultimately, there isn't a whole lot of flexibility in service relationships. It's either messy dependencies, weak ordering schemes, preconditions/postconditions (of which conflicts are a subset), or conditional execution based on resource watching. More esoterically, there's dynamic tracing, but that's outside our scope.

I suppose in SystemXVI's case, an extension of rcorder(8) coupled with a decent framework for either resource monitoring or pre/postconditions (I actually think David wants to do some Upstart-like event thing) would be sufficient.

**Comment by JX7P** [comment]
In terms of the event model, I was thinking more in terms of the architecture of the service ordering daemon. The service ordering daemon would register with repositoryd its interest in service state-change events (which are emitted by the service' restarter) and its event loop would be driven by these events.

On the resource/service concept - consider the case of a service that needs *any* mail routing daemon to be available. We will want to express somehow this relationship in a way that doesn't create a dependency on `sendmail` or `qmail` or any other provider; instead, we want to have a more abstract dependency.

I quite like the idea of a `provides` and `requires` relationship here. We may have a service, mail/sendmail, which provides `mail` and `mail/sendmail`.  Alternatively, a service mail/qmail which provides `mail` and `mail/qmail`. Another service may require either `mail` or some specialisation thereof according to its own needs.

**Comment by uselessd** [comment]
Hm, the event loop based on repository state change notifications is more fine-grained than the clunky Upstart events I recall you proposing earlier. All the logic then goes in the loop, where orderings can be calculated dynamically in response to queued events.

The resource concept, of course, would mean you would adjust the watcher to look for whatever data your current mail router exports. I suppose the high-level provides/requires relationship as seen in simpleinit-msb and similar would be fine, though.

**Comment by JX7P** [comment]
@mkfifo Are you still interested in working on this? Let me know if you are, and if there's any way I can make it easier.
